File: day8/main.go
package main

import (
	"fmt"
	"os"
)

type Node struct {
	MetaNum   int
	Metas     []int
	MetaValue int

	ChildNum int
	Children []*Node
}

func (n *Node) String() string {
	return fmt.Sprintf("[meta_num=%d meta=%d child_num=%d children=%s]", n.MetaNum, n.Metas, n.ChildNum, n.Children)
}

// PraseTree returns the root node of a tree and the remaining input
func ParseTree(input []int) (*Node, []int) {
	// construct node
	n := &Node{}
	n.ChildNum = input[0]
	n.Children = make([]*Node, input[0])
	n.MetaNum = input[1]
	n.Metas = make([]int, input[1])

	// parse children
	remaining := input[2:]
	var child *Node
	for i := range n.Children {
		child, remaining = ParseTree(remaining)
		n.Children[i] = child
	}

	// parse metadata
	for i := range n.Metas {
		n.Metas[i] = remaining[i]
	}
	remaining = remaining[len(n.Metas):]
	return n, remaining
}

// Traverse the tree
func Traverse(root *Node, post func(n *Node)) {
	for _, child := range root.Children {
		Traverse(child, post)
	}
	post(root)
}

func main() {
	// parse
	f, _ := os.Open("input.txt")
	result := make([]int, 0)
	for {
		var i int
		_, err := fmt.Fscan(f, &i)
		if err != nil {
			break
		}
		result = append(result, i)
	}
	// result = []int{2, 3, 0, 3, 10, 11, 12, 1, 1, 0, 1, 99, 2, 1, 1, 2}
	root, _ := ParseTree(result)

	// part 1
	metaSum := 0
	post := func(n *Node) {
		for _, m := range n.Metas {
			metaSum += m
		}
	}
	Traverse(root, post)
	fmt.Println(metaSum)

	// part 2
	post = func(n *Node) {
		var metaVal int
		if n.ChildNum == 0 {
			for _, m := range n.Metas {
				metaVal += m
			}
		} else {
			for _, m := range n.Metas {
				i := m - 1
				if 0 <= i && i < n.ChildNum {  // check index is valid
					child := n.Children[i]
					metaVal += child.MetaValue
				}
			}
		}
		n.MetaValue = metaVal
	}
	Traverse(root, post)
	fmt.Println(root.MetaValue)
}

File: day8/main.py
#!/usr/bin/env python3

with open("input.txt") as f:
    ipt = [int(i) for i in
           f.read().strip().split(' ')]

# Node(child_num: int, meta_num: int, children: []node, metas: []int, meta_val: []int)

def parse_node(l): # => (node, remaining)
    node = (l[0], l[1], [], [], [])

    # parse children
    remaining = l[2:]
    for i in range(node[0]):
        result = parse_node(remaining)
        node[2].append(result[0])
        remaining = result[1]

    # parse metadata
    for i in range(node[1]):
        node[3].append(remaining[0])
        remaining = remaining[1:]

    return node, remaining


def visit(node):
    for i in range(node[0]):
        yield from visit(node[2][i])
    yield node

result = parse_node(ipt)
root = result[0]

# part 1
print(sum(([sum(n[3]) for n in visit(root)])))

# part 2
for n in visit(root):
    if n[0] == 0:
        n[4].append(sum(n[3]))
    else:
        idx = [i - 1 for i in n[3] if i > 0 and i <= n[0]]
        children = [n[2][i] for i in idx]
        n[4].append(sum([child[4][0] for child in children]))

print(root[4][0])
